droidlet/tools/hitl/nsp_retrain_infra.py

```diff
@@ -0,0 +1,370 @@
+"""
+Copyright (c) Facebook, Inc. and its affiliates.
+"""
+
+import argparse
+import logging
+import os
+import shutil
+import time
+from subprocess import Popen, PIPE, TimeoutExpired
+import boto3
+from datetime import datetime
+import torch
+
+from typing import List
+
+from utils.data_generator import DataGenerator
+from utils.job_listener import JobListener
+from utils.task_runner import TaskRunner
+
+
+log_formatter = logging.Formatter(
+    "%(asctime)s [%(filename)s:%(lineno)s - %(funcName)s() %(levelname)s]: %(message)s"
+)
+logger = logging.getLogger()
+logger.handlers.clear()
+logger.setLevel("INFO")
+sh = logging.StreamHandler()
+sh.setFormatter(log_formatter)
+logger.addHandler(sh)
+
+LISTENER_SLEEP_TIME = 10  # Seconds to wait before listener looks for new data
+NSP_RETRAIN_TIMEOUT = 18000  # Wait a max of 5h for the NSP retraining script to finish
+MODEL_OUTPUT_POLL_TIME = 60  # Seconds to wait between looking for model output logs
+NUM_MODEL_OUTPUT_POLLS = 500  # Number of times to poll model output files before assuming finished
+
+s3 = boto3.client('s3')
+
+NUM_TRAINING_RUNS = 6
+MODEL_NAME = "best_model.pth"
+MODEL_INFO_NAME = "best_model_info.txt"
+
+
+class NSPRetrainingJob(DataGenerator):
+    def __init__(self, batch_id, opts):
+        super(NSPRetrainingJob, self).__init__()
+        self.batch_id = batch_id
+        self.opts = opts
+        self.exec_training_run = True
+
+
+    def max_and_argmax(self,l):
+        i = max(range(len(l)), key=lambda i: l[i])
+        return l[i], i
+
+
+    def get_accs(self):
+        accs = {}
+        logs = [l for l in os.listdir() if l[-4:] == ".log"]
+        if len(logs) < NUM_TRAINING_RUNS:  # Wait until all logs have appeared, takes ~30 min from first to last
+            return False
+        for log in logs:
+            accs[log] = []
+            f = open(log)
+            l = f.readlines()
+            for line in l:
+                if "evaluating on" in line:
+                    s = line.split("\t")
+                    a = s[2][11:17]
+                    accs[log].append(float(a))
+        return accs
+
+
+    def copy_best_model(self,accs):
```

Comment:
In rare cases (accuracy == 0.0 for all models), this will result in the following errors: 

```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/private/home/yuxuans/.conda/envs/minecraft_env/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/private/home/yuxuans/.conda/envs/minecraft_env/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "nsp_retrain_infra.py", line 306, in run
    self.copy_best_model(accs)
  File "nsp_retrain_infra.py", line 86, in copy_best_model
    shutil.copyfile(source_model, MODEL_NAME)
  File "/private/home/yuxuans/.conda/envs/minecraft_env/lib/python3.7/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '_ep0.pth'
```

Reply:
We could replace line 80: `if m > macc:` with `if m >= macc:`